video/out/opengl/hwdec_cuda.c

```diff
@@ -197,6 +297,103 @@ static void cuda_uninit(struct ra_hwdec *hw)
 #undef CHECK_CU
 #define CHECK_CU(x) check_cu((mapper)->owner, (x), #x)
 
+#if HAVE_VULKAN
+static struct ra_buf *cuda_buf_pool_get(struct ra_hwdec_mapper *mapper, int n)
+{
+    struct priv_owner *p_owner = mapper->owner->priv;
+    struct priv *p = mapper->priv;
+    CudaFunctions *cu = p_owner->cu;
+    int ret = 0;
+
+    struct ra_buf_pool *pool = &p->buf_pool[n];
+    struct ra_buf *buf = ra_buf_pool_get(mapper->ra, pool, &p->buf_params[n]);
+    if (!buf) {
+        goto error;
+    }
+
+    if (!ra_vk_buf_get_user_data(buf)) {
+        struct ext_buf *ebuf = talloc_zero(NULL, struct ext_buf);
+        struct vk_external_mem mem_info;
+
+        bool success = ra_vk_buf_get_external_info(mapper->ra, buf, &mem_info);
+        if (!success) {
+            ret = -1;
+            goto error;
+        }
+
+#if HAVE_WIN32_DESKTOP
+        MP_DBG(mapper, "vk_external_info[%d][%d]: %p %zu %zu\n", n, pool->index, ebuf->handle, mem_info.size, mem_info.offset);
+        ebuf->handle = mem_info.mem_handle;
+#else
+        MP_DBG(mapper, "vk_external_info[%d][%d]: %d %zu %zu\n", n, pool->index, ebuf->fd, mem_info.size, mem_info.offset);
+        ebuf->fd = mem_info.mem_fd;
+#endif
+
+        CUDA_EXTERNAL_MEMORY_HANDLE_DESC ext_desc = {
+#if HAVE_WIN32_DESKTOP
+            .type = IsWindows8OrGreater()
+                ? CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32
+                : CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_KMT,
+            .handle.win32.handle = ebuf->handle,
+#else
+            .type = CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD,
+            .handle.fd = ebuf->fd,
+#endif
+            .size = mem_info.mem_size,
+            .flags = 0,
+        };
+        ret = CHECK_CU(cu->cuImportExternalMemory(&ebuf->mem, &ext_desc));
+        if (ret < 0)
+            goto error;
+
+        CUDA_EXTERNAL_MEMORY_BUFFER_DESC buf_desc = {
+            .offset = mem_info.offset,
+            .size = mem_info.size,
+            .flags = 0,
+        };
+        ret = CHECK_CU(cu->cuExternalMemoryGetMappedBuffer(&ebuf->buf, ebuf->mem, &buf_desc));
+        if (ret < 0)
+            goto error;
+
+        ra_vk_buf_set_user_data(buf, ebuf);
+    }
+    return buf;
+
+error:
+    MP_ERR(mapper, "cuda_buf_pool_get failed\n");
+    return NULL;
+}
+
+static void cuda_buf_pool_uninit(struct ra_hwdec_mapper *mapper, int n)
+{
+    struct priv_owner *p_owner = mapper->owner->priv;
+    struct priv *p = mapper->priv;
+    CudaFunctions *cu = p_owner->cu;
+
+    struct ra_buf_pool *pool = &p->buf_pool[n];
+    for (int i = 0; i < pool->num_buffers; i++) {
+        struct ra_buf *buf = pool->buffers[i];
+        struct ext_buf *ebuf = ra_vk_buf_get_user_data(buf);
+        if (ebuf) {
+            if (ebuf->mem > 0) {
```

Comment:
This seems sort of weird - are you sure that cuDestroyExternalMemory also implies closing the fd/handle?

Reply:
I had to read the docs really carefully. And even better, it's different for fds vs HANDLEs...

https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html#group__CUDA__EXTRES__INTEROP_1g52aba3a7f780157d8ba12972b2481735

So, for an fd, ownership is fully transferred. For a handle, it is not. Thanks nvidia. Will adjust.